--- ./build/buildroot.regigraf.cortex_a8/package/directfb/directfb.mk.orig	2012-05-31 01:23:07.000000000 +0400
+++ ./build/buildroot.regigraf.cortex_a8/package/directfb/directfb.mk	2013-04-12 20:50:03.000000000 +0400
@@ -21,7 +21,8 @@
 	--disable-osx \
 	--disable-video4linux \
 	--disable-video4linux2 \
-	--without-tools
+	--without-tools \
+	--enable-neon
 
 DIRECTFB_DEPENDENCIES = freetype zlib
 
--- ./build/buildroot.regigraf.cortex_a8/package/libdrm/libdrm.mk.orig	2012-05-31 01:23:07.000000000 +0400
+++ ./build/buildroot.regigraf.cortex_a8/package/libdrm/libdrm.mk	2013-06-22 15:10:26.000000000 +0400
@@ -17,6 +17,8 @@
 	pthread-stubs \
 	host-pkg-config
 
+LIBDRM_MAKE_OPT += CFLAGS="$(TARGET_CFLAGS) -lrt"
+
 ifeq ($(BR2_PACKAGE_XDRIVER_XF86_VIDEO_INTEL),y)
 LIBDRM_CONF_OPT += --enable-intel
 LIBDRM_DEPENDENCIES += libatomic_ops
--- ./build/buildroot.regigraf.cortex_a8/package/webkit/webkit.mk.orig	2012-05-31 01:23:07.000000000 +0400
+++ ./build/buildroot.regigraf.cortex_a8/package/webkit/webkit.mk	2013-06-22 21:48:48.000000000 +0400
@@ -5,7 +5,7 @@
 #############################################################
 
 WEBKIT_VERSION = 1.2.7
-WEBKIT_SITE = http://www.webkitgtk.org
+WEBKIT_SITE = http://www.webkitgtk.org/releases
 WEBKIT_INSTALL_STAGING = YES
 WEBKIT_DEPENDENCIES = host-flex host-gperf icu libcurl libxml2 libxslt \
 			libgtk2 sqlite enchant libsoup jpeg libgail
