--- ./build/linux-2.6.35.3.SK/arch/arm/mach-mx5/mx53_loco.c.orig	2013-06-24 10:37:02.000000000 +0400
+++ ./build/linux-2.6.35.3.SK/arch/arm/mach-mx5/mx53_loco.c	2013-08-21 15:45:14.591844757 +0400
@@ -860,15 +861,17 @@
  * 16M should cover all kind of NAND boot support on MX53.
  */
 	{
-//	 .name   = "bootloader and kernel",
 	 .name   = "bootloader",
 	 .offset = 0,
-   .size   = 16 * 1024 * 1024
-//	 .size = 32 * 1024 * 1024
-
+	 .size   = 0x000100000
+	},
+	{
+	 .name   = "config",
+	 .offset = 0x000100000,
+	 .size   = 0x000100000
 	}, {
 	 .name   = "kernel",
-	 .offset = 16 * 1024 * 1024,
+	 .offset = 0x000200000,
    .size   = 16 * 1024 * 1024
 	}, {
 	 .name   = "rootfs",
--- ./build/linux-2.6.35.3.SK/drivers/video/Kconfig.orig	2011-11-10 01:51:30.000000000 +0400
+++ ./build/linux-2.6.35.3.SK/drivers/video/Kconfig	2013-07-03 15:38:54.867572961 +0400
@@ -207,7 +207,7 @@
        default n
 
 config FB_BACKLIGHT
-	bool
+	bool "Enable backlight control for FB"
 	depends on FB
 	select BACKLIGHT_LCD_SUPPORT
 	select BACKLIGHT_CLASS_DEVICE
 	
